runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1651,6 +1668,261 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
+                                                                  fieldEntry._isPrivate,
+                                                                  fieldEntry._fieldname,
+                                                                  fieldEntry._typeSignature);
+
+      TR::DataType dataType = fieldEntry._datatype;
+      bool useReadBarrier = comp()->getOption(TR_EnableFieldWatch);
+      bool useWriteBarrier = ((dataType == TR::Address) && (TR::Compiler->om.writeBarrierType() != gc_modron_wrtbar_none)) || comp()->getOption(TR_EnableFieldWatch);
+
+      TR::ILOpCodes loadOpCode = useReadBarrier ? comp()->il.opCodeForIndirectReadBarrier(dataType) : comp()->il.opCodeForIndirectLoad(dataType);
+      TR::ILOpCodes storeOpCode = useWriteBarrier ? comp()->il.opCodeForIndirectWriteBarrier(dataType) : comp()->il.opCodeForIndirectStore(dataType);
+
+      if (trace())
+         {
+         traceMsg(comp(),"%s fieldEntry[%d] fieldName %s fieldSig %s type %d offset %d\n", __FUNCTION__,
+            (int)idx, fieldEntry._fieldname, fieldEntry._typeSignature, dataType.getDataType(), fieldEntry._offset);
+         traceMsg(comp(),"    %s loadFieldSymRef %p %s . %s storeFieldSymRef %p %s\n",
+            comp()->getDebug()->getName(loadOpCode), loadFieldSymRef, comp()->getDebug()->getName(loadFieldSymRef),
+            comp()->getDebug()->getName(storeOpCode), storeFieldSymRef, comp()->getDebug()->getName(storeFieldSymRef));
+         }
+
+      TR::Node *loadNode = TR::Node::createWithSymRef(loadOpCode, 1, 1, valueNode, loadFieldSymRef);
+      TR::Node *storeNode = NULL;
+
+      if (idx == 0)
+         {
+         if (useWriteBarrier)
+            storeNode = TR::Node::recreateWithoutProperties(callNode, storeOpCode, 3, elementAddressNode, loadNode, arrayRefNode, storeFieldSymRef);
+         else
+            storeNode = TR::Node::recreateWithoutProperties(callNode, storeOpCode, 2, elementAddressNode, loadNode, storeFieldSymRef);
+         }
+      else
+         {
+         if (useWriteBarrier)
+            storeNode = TR::Node::createWithSymRef(storeOpCode, 3, 3, elementAddressNode, loadNode, arrayRefNode, storeFieldSymRef);
+         else
+            storeNode = TR::Node::createWithSymRef(storeOpCode, 2, 2, elementAddressNode, loadNode, storeFieldSymRef);
+
+         tt = tt->insertAfter(TR::TreeTop::create(comp(), storeNode));
+         }
+      }
+
+   return false;
+   }
+
+TR_YesNoMaybe
+J9::ValuePropagation::isArrayElementFlattened(TR::VPConstraint *arrayConstraint)
+   {
+   if (!TR::Compiler->om.areValueTypesEnabled())
+      {
+      return TR_no;
+      }
+
+   if (!arrayConstraint || !arrayConstraint->getClass())
+      {
+      return TR_maybe;
+      }
+
+   TR_OpaqueClassBlock *arrayClass = arrayConstraint->getClass();
+
+   if ((arrayConstraint->getClassType()->isArray() == TR_yes) &&
+       TR::Compiler->cls.isValueTypeClassFlattened(arrayClass))
+      {
+      return TR_yes;
+      }
+
+   return TR_no;
```

Comment:
>It might be simplest to call J9::ValuePropagation::isArrayCompTypeValueType, 

That's a good idea!

Just to clarify If `arrayConstraint->getClassType()->isArray()` returns `TR_maybe` or `ValuePropagation::isArrayElementFlattened` returns `TR_maybe`, we don't transform the helper calls with sym refs.

Reply:
> Just to clarify If `arrayConstraint->getClassType()->isArray()` returns `TR_maybe` or `ValuePropagation::isArrayElementFlattened` returns `TR_maybe`, we don't transform the helper calls

That's correct

Reply:
Addressed in e5456bc